students/dave_arasim/session05/mailroom3.py

```diff
@@ -0,0 +1,251 @@
+    with open('donor_dict.txt', 'w') as dictfile:
+        for this_name, this_amts in donor_dict.items():
+            amts_str = ','.join(str(x) for x in this_amts) # Comprehension
+            this_line = this_name + ':' + amts_str + '\n'
+
+            dictfile.write(this_line)
+
+    dictfile.close()
+
+
+def send_thank_you(donor_dict):
+    '''Sends thank you letter to single donor as their donation is made and recorded'''
+    quit_thank_you = False
+
+    clear_screen()
+
+    while not(quit_thank_you):
+        print("Enter Donor's name, type 'list' for existing Donors (<cr> to exit): ", end='')
+
+        thank_you_choice = input()
+        print()
+
+        if thank_you_choice == '': quit_thank_you = True
+        elif thank_you_choice == 'list': show_donor_list(donor_dict)
+        else:
+            find_donor(donor_dict, thank_you_choice)
+            quit_thank_you = True
+
+
+def show_donor_list(donor_dict):
+    '''Show the Donor List of names only'''
+    donor_list = 'List of Donors'
+    donor_list_w = 26
+
+    # Donor list header
+    list_nom = (donor_list, donor_list_w)
+    print('{:{}}'.format(*list_nom))
+    print('-'*donor_list_w)
+    
+    # Donor list body
+    for this_name in donor_dict:
+        # Data line
+        this_line = (this_name, donor_list_w)
+        print('{:{}}'.format(*this_line))
+
+    print()
+
+
+def find_donor(donor_dict, thank_you_choice):
+    '''Finds an existing Donor OR installs a new Donor and records donation amount.
+    It then sends a thank you letter to that Donor for the current donation.'''
+    global form_text_str
+
+    valid_amt = False
+
+    while not(valid_amt):
+        print('Please enter donation amount for ', thank_you_choice, ': $', sep='', end='')
+        donation_amt = input()
+
+        try:
+            donation_amt = float(donation_amt)
+        except ValueError:
+            print('Sorry but that is not a valid amount')
+            print()
+        else:
+            valid_amt = True
+
+    donor_dict[thank_you_choice].append(donation_amt)
+
+    form_dict = {'form_name': thank_you_choice, 'form_amt': donation_amt}
+
+    print()
+    print('E-mail form letter output:')
+    print()
+    print(form_text_str.format(**form_dict))
+    print()
+    print('End e-mail form letter output, <cr> to continue... ', end='')
+    cont_input = input()  # cont_input for pause (variable not used)
+
+
+def send_all_thanks(donor_dict):
+    '''Send thank you letters to all donors as text files named as donor_name.txt'''
+    global form_text_str
+
+    for this_name, this_amts in donor_dict.items():
+        thank_you_text = ''
+
+        this_tot = sum(this_amts)
+        this_file = this_name.replace(' ', '_') + '.txt'
+
+        form_dict = {'form_name': this_name, 'form_amt': this_tot}
+
+        thank_you_text = form_text_str.format(**form_dict)
+
+        with open(this_file, 'w') as thank_you_file:
+            thank_you_file.write(thank_you_text)
```

Comment:
Another good place for `try`-`catch` in case the file writing fails due to errors like `PermissionError` or `IOError`.